pkg/app/api/grpcapi/piped_api.go

```diff
@@ -58,13 +58,15 @@ type PipedAPI struct {
 	deploymentPipedCache cache.Cache
 	envProjectCache      cache.Cache
 	pipedStatCache       cache.Cache
+	// A map from projectID to map["repo-id"][]*model.ApplicationInfo
+	unregisteredAppsCache cache.Cache
```

Comment:
Grouping by project ID like this could cause a problem where data is overridden between multiple Pipeds of the same project. (a kind of race condition).
For example, Piped-1 is watching Repo-1, Piped-2 is watching Repo-2, Piped-1 and Piped-2 send their list to the control plane for the same project. So in some conditions, Piped-1's data will be overridden by Piped'2 data.

Besides that, I think the data structure `map["repo-id"][]*model.ApplicationInfo`  inside each project key could not work well since multiple Pipeds can watch the same repository and all of them are sending an `unregistered` list to the control plane.

Reply:
That's the case. Let me re-think what structure is for us.